src/app/workflow/view/view.service.ts

```diff
@@ -0,0 +1,174 @@
+/*
+ *    Copyright 2017 OICR
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+import { Injectable } from '@angular/core';
+import { HttpErrorResponse } from '@angular/common/http';
+import { MatDialog } from '@angular/material';
+import { BehaviorSubject, Observable, of as observableOf, Subject } from 'rxjs';
+import { concatMap } from 'rxjs/operators';
+import { AlertService } from '../../shared/alert/state/alert.service';
+import { RefreshService } from '../../shared/refresh.service';
+import { WorkflowQuery } from '../../shared/state/workflow.query';
+import { WorkflowsService } from '../../shared/swagger/api/workflows.service';
+import { SourceFile } from '../../shared/swagger/model/sourceFile';
+import { WorkflowService } from '../../shared/state/workflow.service';
+import { Workflow, WorkflowVersion } from '../../shared/swagger';
+import { ConfirmationDialogService } from '../../confirmation-dialog/confirmation-dialog.service';
+import { ConfirmationDialogData } from '../../confirmation-dialog/confirmation-dialog.component';
+
+@Injectable()
+export class ViewService {
+  version: Subject<WorkflowVersion> = new BehaviorSubject<WorkflowVersion>(null);
+  constructor(
+    private alertService: AlertService,
+    private confirmationDialogService: ConfirmationDialogService,
+    private workflowQuery: WorkflowQuery,
+    private workflowService: WorkflowService,
+    private workflowsService: WorkflowsService,
+    private refreshService: RefreshService,
+    private matDialog: MatDialog
+  ) {}
+
+  /**
+   * Updates the workflow version and alerts the Dockstore User with success
+   * or failure.
+   *
+   * @private
+   * @memberof ViewService
+   */
+  private updateWorkflowToSnapshot(workflow: Workflow, version: WorkflowVersion, cb: Function): void {
```

Comment:
The callback function is node.js-y. Should use Observable.

Reply:
I left in one way of doing this using observables for the snapshot. I'm not very happy with the refactor it made though. The callback method is still used for the DOI button for comparison (and to get advice).